File: backend/internal/agents/runner/coder_confirmation.go
package runner

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/google/uuid"
)

// ---------- Coder 确认流 ----------

// CoderConfirmPreview 包含用于前端预览的工具调用详情。
type CoderConfirmPreview struct {
	FilePath  string `json:"filePath,omitempty"`
	OldString string `json:"oldString,omitempty"` // edit: 待替换文本片段
	NewString string `json:"newString,omitempty"` // edit: 新文本片段
	Content   string `json:"content,omitempty"`   // write: 内容预览 (截断 500 字符)
	Command   string `json:"command,omitempty"`   // bash: 命令文本
	LineCount int    `json:"lineCount,omitempty"` // write: 内容行数
}

// CoderConfirmationRequest 表示一个等待用户确认的 coder 操作。
type CoderConfirmationRequest struct {
	ID          string               `json:"id"`
	ToolName    string               `json:"toolName"` // "edit"|"write"|"bash"
	Args        json.RawMessage      `json:"args"`     // 原始参数
	Preview     *CoderConfirmPreview `json:"preview"`  // 预览数据
	CreatedAtMs int64                `json:"createdAtMs"`
	ExpiresAtMs int64                `json:"expiresAtMs"`
}

// CoderConfirmBroadcastFunc 广播回调（解耦 runner 与 gateway）。
type CoderConfirmBroadcastFunc func(event string, payload interface{})

// CoderConfirmationManager 管理 coder 工具调用确认流。
// 当 coder 触发可确认工具 (edit/write/bash) 时：
//  1. 广播 "coder.confirm.requested" 给前端
//  2. 阻塞等待用户决策（allow/deny）或超时
//  3. 前端通过 "coder.confirm.resolve" RPC 回调
//
// 为 nil 时完全跳过确认（兼容现有行为）。
type CoderConfirmationManager struct {
	mu        sync.Mutex
	pending   map[string]chan string // id → decision ("allow"/"deny")
	broadcast CoderConfirmBroadcastFunc
	timeout   time.Duration // 默认 60s
}

// NewCoderConfirmationManager 创建确认管理器。
func NewCoderConfirmationManager(broadcastFn CoderConfirmBroadcastFunc, timeout time.Duration) *CoderConfirmationManager {
	if timeout <= 0 {
		timeout = 60 * time.Second
	}
	return &CoderConfirmationManager{
		pending:   make(map[string]chan string),
		broadcast: broadcastFn,
		timeout:   timeout,
	}
}

// RequestConfirmation 请求用户确认一个 coder 操作。
// 阻塞直到用户决策、超时或 ctx 取消。
// 返回 true 表示用户批准，false 表示拒绝/超时/取消。
func (m *CoderConfirmationManager) RequestConfirmation(ctx context.Context, toolName string, args json.RawMessage) (bool, error) {
	now := time.Now()
	req := CoderConfirmationRequest{
		ID:          uuid.New().String(),
		ToolName:    toolName,
		Args:        args,
		Preview:     extractCoderPreview(toolName, args),
		CreatedAtMs: now.UnixMilli(),
		ExpiresAtMs: now.Add(m.timeout).UnixMilli(),
	}

	ch := make(chan string, 1)

	m.mu.Lock()
	m.pending[req.ID] = ch
	m.mu.Unlock()

	// 广播确认请求到前端
	if m.broadcast != nil {
		m.broadcast("coder.confirm.requested", req)
	}

	slog.Debug("coder confirmation requested",
		"id", req.ID,
		"tool", toolName,
	)

	// 等待用户决策、超时或 ctx 取消
	timer := time.NewTimer(m.timeout)
	defer timer.Stop()

	var decision string
	select {
	case decision = <-ch:
		// 用户已决策
	case <-timer.C:
		decision = "deny"
		slog.Info("coder confirmation timed out, auto-denying",
			"id", req.ID,
			"tool", toolName,
		)
	case <-ctx.Done():
		decision = "deny"
		slog.Debug("coder confirmation cancelled by context",
			"id", req.ID,
		)
	}

	// 清理 pending
	m.mu.Lock()
	delete(m.pending, req.ID)
	m.mu.Unlock()

	// 广播决策结果
	if m.broadcast != nil {
		m.broadcast("coder.confirm.resolved", map[string]interface{}{
			"id":       req.ID,
			"decision": decision,
			"ts":       time.Now().UnixMilli(),
		})
	}

	return decision == "allow", nil
}

// ResolveConfirmation 处理前端的确认决策回调。
// 由 WebSocket RPC "coder.confirm.resolve" 调用。
func (m *CoderConfirmationManager) ResolveConfirmation(id, decision string) error {
	if decision != "allow" && decision != "deny" {
		return fmt.Errorf("invalid decision: %q (expected allow/deny)", decision)
	}

	m.mu.Lock()
	ch, ok := m.pending[id]
	m.mu.Unlock()

	if !ok {
		return fmt.Errorf("no pending confirmation with id: %s", id)
	}

	// 非阻塞写入（channel 有 1 缓冲）
	select {
	case ch <- decision:
		slog.Debug("coder confirmation resolved",
			"id", id,
			"decision", decision,
		)
	default:
		// channel 已被写入（超时或重复调用），忽略
	}

	return nil
}

// isCoderConfirmable 判断工具是否需要确认。
func isCoderConfirmable(mcpToolName string) bool {
	switch mcpToolName {
	case "edit", "write", "bash":
		return true
	default:
		return false
	}
}

// extractCoderPreview 从工具参数中提取预览数据。
func extractCoderPreview(toolName string, args json.RawMessage) *CoderConfirmPreview {
	var parsed map[string]interface{}
	if err := json.Unmarshal(args, &parsed); err != nil {
		return nil
	}

	preview := &CoderConfirmPreview{}

	switch toolName {
	case "edit":
		if v, ok := parsed["filePath"].(string); ok {
			preview.FilePath = v
		}
		if v, ok := parsed["oldString"].(string); ok {
			preview.OldString = truncatePreview(v, 500)
		}
		if v, ok := parsed["newString"].(string); ok {
			preview.NewString = truncatePreview(v, 500)
		}
	case "write":
		if v, ok := parsed["filePath"].(string); ok {
			preview.FilePath = v
		}
		if v, ok := parsed["content"].(string); ok {
			preview.Content = truncatePreview(v, 500)
			preview.LineCount = countLines(v)
		}
	case "bash":
		if v, ok := parsed["command"].(string); ok {
			preview.Command = v
		}
	}

	return preview
}

// truncatePreview 截断预览文本到指定长度。
func truncatePreview(s string, maxLen int) string {
	runes := []rune(s)
	if len(runes) <= maxLen {
		return s
	}
	return string(runes[:maxLen]) + "..."
}

// countLines 统计文本行数。
func countLines(s string) int {
	if s == "" {
		return 0
	}
	count := 1
	for _, c := range s {
		if c == '\n' {
			count++
		}
	}
	return count
}

File: ui/src/ui/controllers/coder-confirmation.ts
// ---------- Coder 确认流控制器 ----------
// 复用 exec-approval.ts 的模式：解析 WebSocket 事件 → 队列管理 → 过期清理。

export type CoderConfirmPreview = {
  filePath?: string;
  oldString?: string;
  newString?: string;
  content?: string;
  command?: string;
  lineCount?: number;
};

export type CoderConfirmRequest = {
  id: string;
  toolName: string; // "edit" | "write" | "bash"
  preview: CoderConfirmPreview;
  createdAtMs: number;
  expiresAtMs: number;
};

export type CoderConfirmResolved = {
  id: string;
  decision: string; // "allow" | "deny"
  ts?: number | null;
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

/**
 * 解析 "coder.confirm.requested" WebSocket 事件 payload。
 */
export function parseCoderConfirmRequested(payload: unknown): CoderConfirmRequest | null {
  if (!isRecord(payload)) return null;

  const id = typeof payload.id === "string" ? payload.id.trim() : "";
  const toolName = typeof payload.toolName === "string" ? payload.toolName.trim() : "";
  if (!id || !toolName) return null;

  const createdAtMs = typeof payload.createdAtMs === "number" ? payload.createdAtMs : 0;
  const expiresAtMs = typeof payload.expiresAtMs === "number" ? payload.expiresAtMs : 0;
  if (!createdAtMs || !expiresAtMs) return null;

  const rawPreview = isRecord(payload.preview) ? payload.preview : {};
  const preview: CoderConfirmPreview = {
    filePath: typeof rawPreview.filePath === "string" ? rawPreview.filePath : undefined,
    oldString: typeof rawPreview.oldString === "string" ? rawPreview.oldString : undefined,
    newString: typeof rawPreview.newString === "string" ? rawPreview.newString : undefined,
    content: typeof rawPreview.content === "string" ? rawPreview.content : undefined,
    command: typeof rawPreview.command === "string" ? rawPreview.command : undefined,
    lineCount: typeof rawPreview.lineCount === "number" ? rawPreview.lineCount : undefined,
  };

  return { id, toolName, preview, createdAtMs, expiresAtMs };
}

/**
 * 解析 "coder.confirm.resolved" WebSocket 事件 payload。
 */
export function parseCoderConfirmResolved(payload: unknown): CoderConfirmResolved | null {
  if (!isRecord(payload)) return null;
  const id = typeof payload.id === "string" ? payload.id.trim() : "";
  if (!id) return null;
  return {
    id,
    decision: typeof payload.decision === "string" ? payload.decision : "deny",
    ts: typeof payload.ts === "number" ? payload.ts : null,
  };
}

function pruneQueue(queue: CoderConfirmRequest[]): CoderConfirmRequest[] {
  const now = Date.now();
  return queue.filter((entry) => entry.expiresAtMs > now);
}

export function addCoderConfirm(
  queue: CoderConfirmRequest[],
  entry: CoderConfirmRequest,
): CoderConfirmRequest[] {
  const next = pruneQueue(queue).filter((item) => item.id !== entry.id);
  next.push(entry);
  return next;
}

export function removeCoderConfirm(
  queue: CoderConfirmRequest[],
  id: string,
): CoderConfirmRequest[] {
  return pruneQueue(queue).filter((entry) => entry.id !== id);
}
